theano/gpuarray/opt.py

```diff
@@ -250,53 +260,69 @@ def apply(self, fgraph):
         # Iterating through inputs of graph
         for i in fgraph.inputs:
             if isinstance(i.type, tensor.TensorType):
-                mapping[i] = GpuFromHost(None)(i)
+                mapping[i] = as_gpuarray_variable(i, None)  # TODO context
             else:
                 mapping[i] = i
-
-        # Iterating through output of all the nodes
-        for n in fgraph.toposort():
-            for o in n.outputs:
-                if isinstance(o.type, tensor.TensorType):
-                    mapping[o] = GpuFromHost(None)(o)
-                else:
-                    mapping[o] = o
+        for i in fgraph.variables:
+            if isinstance(i, theano.Constant):
+                mapping[i] = i
 
         for node in fgraph.toposort():
 
-            # The Extra condition
-            if any([isinstance(i, GpuArrayVariable) or
-               isinstance(i, GpuArraySharedVariable)
-               for i in node.inputs + node.outputs]):
+            if isinstance(node.op, HostFromGpu):
+                mapping[node.outputs[0]] = node.inputs[0]
```

Comment:
Can you do a new timing with the fix above and a second timing without that condition (so that we always more it to the GPU?